### PR TITLE
[jit] Add LSTM to standard library

### PR DESCRIPTION
**Pull request by driazati**
**[WIP]**

Adds support for `torch.nn.LSTM` in Script
* De-sugaring for `self._parameters.values()` to a `Tensor[]`
	* add `aten::List` for `int`, `float`, and `Tensor`
* `LSTM` accepts both `PackedSequence` (which is a `Tuple[Tensor, Optional[Tensor]]`) or `Tensor` for[WI{{ its input. While the Python version is unaffected, the TorchScript `LSTM` only supports `Tuple[Tensor, Optional[Tensor]]` for its input.
	* `aten::_wrap_tuple` and `aten::_unwrap_tuple` are used to provide the correct types to the Script compiler, but cannot actually be run

### CODE REVIEW

File: torch/csrc/jit/script/init.cpp [hunk 1]
```diff
@@ -193,14 +194,70 @@ struct VISIBILITY_HIDDEN ConstantPythonTupleValue : public PythonValue {
...
+  : module_(std::move(module)) {}
+
+  std::string kind() const override {
+    return "parameter list";
+  }
+
+  // self._parameters.values()
+  std::shared_ptr<SugaredValue> call(SourceRange loc, Method & caller, at::ArrayRef<NamedValue> inputs, at::ArrayRef<NamedValue> attributes, size_t n_binders) override {
+    std::vector<Value*> params;
+    const auto& param_list = module_->get_parameters();
```
Comment (wanchaol) [inline torch/csrc/jit/script/init.cpp]: You might also need to check if the parameters is buffer or not

File: torch/jit/__init__.py [hunk 2]
```diff
@@ -1402,6 +1403,14 @@ def _unwrap_optional(x):
     return x
 
 
+def _unwrap_tuple(x):
```
Comment (wanchaol) [inline torch/jit/__init__.py]: hmm do we need to think more about this? I think we want to ultimately remove any those unwrap functions if possible, adding more and more might result in a non-easy revert in the future. At least we should think how to provide user a API or something they don't need to care about.

File: torch/csrc/jit/register_special_ops.cpp [hunk 3]
```diff
@@ -96,6 +106,49 @@ RegisterOperators reg({
...
+          return [](Stack& stack) {
+            auto tuple = pop(stack).toTuple()->elements();
+            JIT_ASSERT(tuple.size() == 2);
+            push(stack, true);
+            return 0;
+          };
+        }),
+    Operator(
+      // "aten::_get_packed_sequence(Tensor a, Tensor b) -> (Tensor, Tensor?)", // TODO: using this causes a segfault
+      FunctionSchema(
```
Comment (wanchaol) [inline torch/csrc/jit/register_special_ops.cpp]: I have the same doubt, why you are using FunctionSchema instead of a string?

File: torch/jit/__init__.py (hunk 2 above)
Comment (driazati) [reply torch/jit/__init__.py]: If we want to support it at the language level instead of having these hacks I would imagine it would look something like

```python
@torch.jit.script
def fn(x):
    # type: (Union[int, float]) -> Union[int, float]
    if isinstance(x, int):
        return 3
    else:
        return 3.5
```

And similarly to #14533 we would only emit the branches for the types seen at compile time

Comment (wanchaol) [reply torch/jit/__init__.py]: Hmmm It's hard to control those meta programming conditions since it might be a runtime known value, it might be ok for us as a temp hack to add unwrap/wrap tuples in supporting lstm modules, I guess we just need to be aware to not stacking too much on them

File: torch/csrc/jit/script/init.cpp [hunk 4]
```diff
@@ -193,14 +194,67 @@ struct VISIBILITY_HIDDEN ConstantPythonTupleValue : public PythonValue {
   }
 };
 
+
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: I doubt it is a good idea to add more SugaredValue types simply to support this one module. This needs a more thorough explanation of what is going on, and why these changes are necessary.

File: torch/nn/modules/rnn.py [hunk 5]
```diff
@@ -437,6 +439,81 @@ class LSTM(RNNBase):
     def __init__(self, *args, **kwargs):
         super(LSTM, self).__init__('LSTM', *args, **kwargs)
 
+    @weak_script_method
```
Comment (zdevito) [inline torch/nn/modules/rnn.py]: This seems like a massive copy-paste from somewhere, what is going on?

**Review by zdevito — COMMENTED** [review]
There needs to be more explanation of what this is trying to accomplish.  There seems to be major copy pasting and adding of specific features to get this to work. There is likely a better way.

**Review by eellison — COMMENTED** [review]
I understand that the issue is the function can be called in python with either a Tensor or a Packed Sequence. Torchscript will only be able to handle the Packed Sequence case. 

I think another approach would be:
```
@torch.jit.script
def fn(x):
    # type: (Tuple[Tensor, Optional[Tensor])
    if isinstance(x, torch.Tensor)
        ... x is typed as Tensor here, and the if branch will be constant prop'd away
    else:
        .... x is a tuple here 
```
This would allow us the torchschript code to compile, without changing the python code.

**Comment by zou3519** [comment]
Btw this will conflict with https://github.com/pytorch/pytorch/pull/15225
